utils/hct/hctdb.py

```diff
@@ -7656,17 +7656,34 @@ def build_valrules(self):
         self.add_valrule(
             "Sm.TGSMUnsupported", "Thread Group Shared Memory not supported %0."
         )
-        self.add_valrule(
+        self.add_valrule_msg(
             "Sm.WaveSizeValue",
-            "Declared WaveSize %0 outside valid range [%1..%2], or not a power of 2.",
+            "WaveSize value must be a power of 2 in range [4..128]",
+            "WaveSize %0 (%1) outside valid range [%2..%3], or not a power of 2.",
         )
-        self.add_valrule(
-            "Sm.WaveSizePreferredOutOfRange",
-            "Preferred WaveSize %0 outside valid range [%1..%2]",
+        self.add_valrule_msg(
+            "Sm.WaveSizeAllZeroWhenUndefined",
+            "WaveSize Max and Preferred must be 0 when Min is 0",
+            "WaveSize Max (%0) and Preferred (%1) must be 0 when Min is 0",
+        )
+        self.add_valrule_msg(
+            "Sm.WaveSizeMaxAndPreferredZeroWhenNoRange",
+            "WaveSize Max and Preferred must be 0 to encode min==max",
+            "WaveSize Max (%0) and Preferred (%1) must be 0 to encode min==max",
+        )
```

Comment:
That is an open issue whether we should prevent the new tag in that case.  I don't think we should prevent the new form for defining any valid WaveSize state because it's simpler.  Currently the validator does not prevent the new form of metadata from being used.

Also, note that the spec proposal was reviewed and approved for getting added, but not all issues were resolved.  So, implementing some things to the letter when decisions haven't been finalized will result in more work in the short run, and may result in more work in the long run as well.  As long as the decisions don't impact things in a way that makes currently valid generated DXIL invalid, we should be good.